src/ProjectSelectionManager.cpp

```diff
@@ -47,88 +47,51 @@ const ProjectSelectionManager &ProjectSelectionManager::Get(
-   auto &settings = ProjectSettings::Get( project );
-   auto &window = ProjectWindow::Get( project );
-   auto snapTo = settings.GetSnapTo();
-   if (snapTo != SNAP_OFF) {
-      auto &viewInfo = ViewInfo::Get( project );
-      auto &selectedRegion = viewInfo.selectedRegion;
-      NumericConverter nc(NumericConverter::TIME,
-         formats.GetSelectionFormat(), 0, ProjectRate::Get(project).GetRate());
-      const bool nearest = (snapTo == SNAP_NEAREST);
-
-      const double oldt0 = selectedRegion.t0();
-      const double oldt1 = selectedRegion.t1();
-
-      nc.ValueToControls(oldt0, nearest);
-      nc.ControlsToValue();
-      const double t0 = nc.GetValue();
-
-      nc.ValueToControls(oldt1, nearest);
-      nc.ControlsToValue();
-      const double t1 = nc.GetValue();
-
-      if (t0 != oldt0 || t1 != oldt1) {
-         selectedRegion.setTimes(t0, t1);
-         return true;
-      }
-   }
+ProjectSelectionManager::ProjectSelectionManager(AudacityProject& project)
+    : mProject { project }
+    , mSnapModeChangedSubscription { ProjectSnap::Get(project).Subscribe(
+         [this](auto&)
+         {
+            SnapSelection();
+         })}
+    , mTimeSignatureChangedSubscription {
+       ProjectTimeSignature::Get(project).Subscribe([this](auto&)
+                                                    { SnapSelection(); })
+    }
 
-   return false;
-}
-
-double ProjectSelectionManager::AS_GetRate()
 {
-   return ProjectRate::Get(mProject).GetRate();
 }
 
-void ProjectSelectionManager::AS_SetRate(double rate)
-{
-   auto &project = mProject;
-   ProjectRate::Get( project ).SetRate( rate );
-}
+ProjectSelectionManager::~ProjectSelectionManager() = default;
 
-int ProjectSelectionManager::AS_GetSnapTo()
-{
-   auto &project = mProject;
-   auto &settings = ProjectSettings::Get( project );
-   return settings.GetSnapTo();
-}
+void ProjectSelectionManager::SnapSelection()
+{   
+   auto& project = mProject;
+   auto& projectSnap = ProjectSnap::Get(mProject);
 
-void ProjectSelectionManager::AS_SetSnapTo(int snap)
-{
-   auto &project = mProject;
-   auto &settings = ProjectSettings::Get( project );
-   auto &window = ProjectWindow::Get( project );
+   if (projectSnap.GetSnapMode() == SnapMode::SNAP_OFF)
+      return;
 
-   settings.SetSnapTo( snap );
+   auto& viewInfo = ViewInfo::Get(project);
+   auto& selectedRegion = viewInfo.selectedRegion;
 
-// LLL: TODO - what should this be changed to???
-// GetCommandManager()->Check(wxT("Snap"), mSnapTo);
-   gPrefs->Write(wxT("/SnapTo"), snap);
-   gPrefs->Flush();
+   const double oldt0 = selectedRegion.t0();
+   const double oldt1 = selectedRegion.t1();
 
-   SnapSelection();
+   const double t0 = projectSnap.SnapTime(oldt0).time;
+   const double t1 = projectSnap.SnapTime(oldt1).time;
 
-   window.RedrawProject();
+   if (t0 != oldt0 || t1 != oldt1)
+   {
+      selectedRegion.setTimes(t0, t1);
+      TrackPanel::Get(mProject).Refresh(false);
```

Comment:
Let this stand for now, but be aware that I'll probably need to break dependencies on TrackPanel and also ProjectWindow to make further progress to toolkit neutrality for larger parts of the code.

TrackPanel should instead either subscribe to change in selection and refresh itself, or poll for changes in idle time.

Reply:
Yeah, I think I will leave it for now.

Reply:
What I described is now part of my own PR, to follow this one